subxt/src/events.rs

```diff
@@ -38,17 +37,357 @@ use codec::{
+        // some event details, then poll the block header subscription to get one.
+        if self.at.is_none() {
+            match futures::ready!(self.block_header_subscription.poll_next_unpin(cx)) {
+                None => {
+                    self.finished = true;
+                    return Poll::Ready(None)
+                }
+                Some(Err(e)) => {
+                    self.finished = true;
+                    return Poll::Ready(Some(Err(e.into())))
+                }
+                Some(Ok(block_header)) => {
+                    use sp_runtime::traits::Header;
+                    // Note [jsdw]: We may be able to get rid of the per-item allocation
+                    // with https://github.com/oblique/reusable-box-future.
+                    self.at = Some(Box::pin(at(self.client, block_header.hash())));
+                    // Continue, so that we poll this function future we've just created.
+                }
+            }
+        }
+
+        // If there is an `at` function stored, we are currently resolving a block header into
+        // some events, so poll that until it's ready.
+        if let Some(res) = &mut self.at {
+            let events = futures::ready!(res.poll_unpin(cx));
+            self.at = None;
+            return Poll::Ready(Some(events))
+        }
+
+        unreachable!()
+    }
+}
+
+/// A collection of events obtained from a block, bundled with the necessary
+/// information needed to decode and iterate over them.
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+pub struct Events<'a, T: Config, Evs: Decode> {
+    metadata: &'a Metadata,
+    block_hash: T::Hash,
+    // Note; raw event bytes are prefixed with a Compact<u32> containing
+    // the number of events to be decoded. We should have stripped that off
+    // before storing the bytes here.
+    event_bytes: Vec<u8>,
+    num_events: u32,
+    _event_type: std::marker::PhantomData<Evs>,
+}
+
+impl<'a, T: Config, Evs: Decode> Events<'a, T, Evs> {
+    /// The number of events.
+    pub fn len(&self) -> u32 {
+        self.num_events
+    }
 
-/// Raw bytes for an Event
-#[derive(Debug)]
-#[cfg_attr(test, derive(PartialEq, Clone))]
-pub struct RawEvent {
+    /// Are there no events in this block?
+    // Note: mainly here to satisfy clippy.
+    pub fn is_empty(&self) -> bool {
+        self.num_events == 0
+    }
+
+    /// Return the block hash that these events are from.
+    pub fn block_hash(&self) -> T::Hash {
+        self.block_hash
+    }
+
+    /// Iterate over the events, statically decoding them as we go.
+    /// If an event is encountered that cannot be statically decoded,
+    /// a [`codec::Error`] will be returned.
+    ///
+    /// If the generated code does not know about all of the pallets that exist
+    /// in the runtime being targeted, it may not know about all of the
+    /// events either, and so this method should be avoided.
```

Comment:
>  if encountering the desired event before any unknown/incompatible events

Actually having read more code of the code I think this is wrong. looks like it will attempt to decode all events in the block first rather than returning early if it finds the one it wants.

Reply:
Definitely worth adding a test for this, but because it decodes one event at a time during iteration, and `find_first_event`/`find` use this iterator, they should succeed up until the first event that cannot be decoded properly, and then return an error and then after that, `None`.